dagutils/traversal.go

```diff
@@ -0,0 +1,507 @@
+	// parent can be set past all of its child nodes (having a value equal to
+	// `len(Links())`) to signal it has visited (or skipped) all of them. A
+	// leaf node with no children and its index in zero would also comply with
+	// this format.
+	childIndex []uint
+	// TODO: Revisit name, `childPosition`? Do not use the *link* term, put
+	// emphasis in the child (the *what*, not the *how* we get it).
+
+	// Depth of the node of the current position. It grows downwards, root is
+	// level 0, its child level 1, and so on. It controls the effective length
+	// of `path` and `childIndex`.
+	//
+	// A level of -1 signals the start case of a new `Traversal` that hasn't
+	// moved yet (or its position has been reset, see `ResetPosition`).
+	// Although this state is an invalid index to the slices, it allows to
+	// centralize all the visit calls in the `down` move (starting at zero
+	// would require a special visit case inside every traversal operation like
+	// `Iterate()` and `Search`). This value should never be returned to after
+	// the first `down` movement, moving up from the root should always return
+	// `ErrUpOnRoot`.
+	level int
+
+	// Method called each time a node is arrived upon in a traversal (through
+	// the `down()` movement, that is, when visiting it for the first time,
+	// not when going back up). It is the main API to implement functionality
+	// (like read and seek) on top of the `Traversal` structure.
+	//
+	// Its argument is the current `node` being visited. Any error it returns
+	// (apart from the internal `errDownStopIteration`) will be forwarded to
+	// the caller of the traversal operation (stopping it).
+	//
+	// Any of the exported methods of this API should be allowed to be called
+	// from within this method.
+	VisitHandler func(node ipld.Node) error
+
+	// Flag to stop the current iteration, intended to be set by the user
+	// inside `VisitHandler` to stop the current traversal operation.
+	Stop bool
+	// TODO: Use a function?
+
+	// Attribute needed to fetch nodes, will become really useful once node
+	// promises are implemented.
+	ctx  context.Context
+	serv ipld.NodeGetter
+
+	// The CID of each child of each node in the `path` (indexed by
+	// `level` and `childIndex`).
+	childCIDs [][]*cid.Cid
+
+	// NodePromises for child nodes requested for every node in the
+	// `path` (as `childCIDs`, indexed by `level` and `childIndex`).
+	promises [][]*ipld.NodePromise
+	// TODO: Consider encapsulating in a single structure along `childCIDs`.
+
+	// Cancel methods for every node in the `path` that had requested
+	// child nodes through the `promises`.
+	cancel []func()
+}
+
+// NewTraversal creates a new `Traversal` structure from a `root`
+// IPLD node.
+func NewTraversal(ctx context.Context, root ipld.Node, serv ipld.NodeGetter) *Traversal {
+	return &Traversal{
+		path:       []ipld.Node{root},
+		childIndex: []uint{0},
+		level:      -1,
+		// Starting position, "on top" of the root node, see `level`.
+
+		ctx:  ctx,
```

Comment:
should avoid storing the context in the struct, it's generally considered a bad practice.

Reply:
This functionality was inherited from the `PBDagReader` which stored the context to later be able to 
 call `ipld.GetNodes`. How would you advise this be handled?